sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferCatalog.scala

```diff
@@ -304,8 +504,196 @@ class RapidsBufferCatalog extends AutoCloseable with Arm with Logging {
     bufferMap.compute(buffer.id, updater)
   }
 
-  /** Remove a buffer ID from the catalog at the specified storage tier. */
-  def removeBufferTier(id: RapidsBufferId, tier: StorageTier): Unit = {
+  /**
+   * Free memory in `store` by spilling buffers to the spill store synchronously.
+   * @param targetTotalSize maximum total size of this store after spilling completes
+   * @return optionally number of bytes that were spilled, or None if this called
+   *         made no attempt to spill due to a detected spill race
+   */
+  def synchronousSpill(store: RapidsBufferStore, targetTotalSize: Long): Option[Long] =
+    synchronousSpill(store, targetTotalSize, Cuda.DEFAULT_STREAM)
+
+  /**
+   * Free memory in `store` by spilling buffers to the spill store synchronously.
+   * @param store store to spill from
+   * @param targetTotalSize maximum total size of this store after spilling completes
+   * @param stream CUDA stream to use or null for default stream
+   * @return optionally number of bytes that were spilled, or None if this called
+   *         made no attempt to spill due to a detected spill race
+   */
+  private def synchronousSpill(
+      store: RapidsBufferStore,
+      targetTotalSize: Long,
+      stream: Cuda.Stream): Option[Long] = {
+    require(targetTotalSize >= 0, s"Negative spill target size: $targetTotalSize")
+    logWarning(s"Trying to spill ${targetTotalSize}. " +
+      s"From store ${store.name}. " +
+      s"Current total ${store.currentSize}. " +
+      s"Current spillable ${store.currentSpillableSize}")
+
+    // we try to spill in this thread. If another thread is also spilling, we let that
+    // thread win and we return letting RMM retry the alloc
+    var rmmShouldRetryAlloc = false
+
+    // total amount spilled in this invocation
+    var totalSpilled: Long = 0
+
+    if (store.currentSpillableSize > targetTotalSize) {
+      withResource(new NvtxRange(s"${store.name} sync spill", NvtxColor.ORANGE)) { _ =>
+        logWarning(s"${store.name} store spilling to reduce usage from " +
+          s"${store.currentSize} total (${store.currentSpillableSize} spillable) " +
+          s"to $targetTotalSize bytes")
+
+        // waited: the store spilled, but it didn't free right away. Instead it waits for
+        // buffers to be closed by clients (since they had references)
+        var waited = false
+
+        // If the store has 0 spillable bytes left, it has exhausted.
+        var exhausted = false
+
+        while (!exhausted && !rmmShouldRetryAlloc &&
+            store.currentSpillableSize > targetTotalSize) {
+          val mySpillCount = spillCount.incrementAndGet()
```

Comment:
- read here, don't increment.
- get lock
- re-read if I am the same one spill, else retry.

That way the first one will win, rather than the last one.

Reply:
Otherwise this could livelock. If the lock is held when incrementing, you guarantee that at least 1 thread has acquired the lock and could spill.